staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/leaselock.go

```diff
@@ -0,0 +1,181 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcelock
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"k8s.io/api/coordination/v1beta1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coordinationv1beta1 "k8s.io/client-go/kubernetes/typed/coordination/v1beta1"
+)
+
+type LeaseLock struct {
+	// LeaseMeta should contain a Name and a Namespace of a
+	// LeaseMeta object that the LeaderElector will attempt to lead.
+	LeaseMeta  metav1.ObjectMeta
+	Client     coordinationv1beta1.LeasesGetter
+	LockConfig ResourceLockConfig
+	l          *v1beta1.Lease
+}
+
+// Get returns the election record from a Lease spec
+func (ll *LeaseLock) Get() (*LeaderElectionRecord, error) {
+	var record LeaderElectionRecord
+	var err error
+	ll.l, err = ll.Client.Leases(ll.LeaseMeta.Namespace).Get(ll.LeaseMeta.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	record, err = leaseSpecToLeaderElectionRecord(&ll.l.Spec)
+	if err != nil {
+		return nil, err
+	}
+	return &record, nil
+}
+
+// Create attempts to create a Lease
+func (ll *LeaseLock) Create(ler LeaderElectionRecord) error {
+	spec, err := LeaderElectionRecordToLeaseSpec(&ler)
+	if err != nil {
+		return err
+	}
+	ll.l, err = ll.Client.Leases(ll.LeaseMeta.Namespace).Create(&v1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ll.LeaseMeta.Name,
+			Namespace: ll.LeaseMeta.Namespace,
+		},
+		Spec: spec,
+	})
+	return err
+}
+
+// Update will update an existing Lease spec.
+func (ll *LeaseLock) Update(ler LeaderElectionRecord) error {
+	if ll.l == nil {
+		return errors.New("lease not initialized, call get or create first")
+	}
+	spec, err := LeaderElectionRecordToLeaseSpec(&ler)
+	if err != nil {
+		return err
+	}
+	ll.l.Spec = spec
+	ll.l, err = ll.Client.Leases(ll.LeaseMeta.Namespace).Update(ll.l)
+	return err
+}
+
+// RecordEvent in leader election while adding meta-data
+func (ll *LeaseLock) RecordEvent(s string) {
+	events := fmt.Sprintf("%v %v", ll.LockConfig.Identity, s)
+	ll.LockConfig.EventRecorder.Eventf(&v1beta1.Lease{ObjectMeta: ll.l.ObjectMeta}, v1.EventTypeNormal, "LeaderElection", events)
+}
+
+// Describe is used to convert details on current resource lock
+// into a string
+func (ll *LeaseLock) Describe() string {
+	return fmt.Sprintf("%v/%v", ll.LeaseMeta.Namespace, ll.LeaseMeta.Name)
+}
+
+// returns the Identity of the lock
+func (ll *LeaseLock) Identity() string {
+	return ll.LockConfig.Identity
+}
+
+func leaseSpecToLeaderElectionRecord(spec *v1beta1.LeaseSpec) (LeaderElectionRecord, error) {
+	holderIdentity := ""
+	if spec.HolderIdentity != nil {
+		holderIdentity = *spec.HolderIdentity
+	}
+	leaseDurationSeconds := 0
```

Comment:
this allows empty holder identity and zero lease duration in the record, which looks slightly odd. but stepping thru the elector, it will work fine i suppose.